### PR TITLE
pkg: Add detection for HA mode based on Infrastructure Topology

### PR DESCRIPTION
**Pull request by lilic**
@simonpasquier I think this covers everything it based on the proposal doc you shared. No tests were added as we do not have a single node env.

I tried launching a cluster with single node but topology was always HA. Tested manually by expecting the reverse, Prometheus scaled down to 1 instance and stack remained in operation. 

cc @openshift/openshift-team-monitoring please have a look.

* [ ] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

**Comment by s-urbaniak** [comment]
lgtm from my side :+1:

**Comment by lilic** [comment]
/retest

Unrelated failure I believe.

**Comment by lilic** [comment]
Yes that would work as well, why I did it in the reconcile tasks was this could be reused in the same way later on if we ever want to expose this configuration to our users. For example if we expose a field in the configuration to allow sharding of Prometheus, the number of shards would need to be passed in configmap level or CR level.

Another approach why I took it this way is to reconcile based on the Infrastructure field, I assumed it could change if topology changes, I think its out of scope now but we might not be in the future.

I will look if this all works for the NewDeployment or if there are any drawbacks. Let me know your feedback based on my reasoning as well, thanks!

**Comment by paulfantom** [comment]
/hold

additionally holding to wait for #1044

**Comment by simonpasquier** [comment]
> this could be reused in the same way later on if we ever want to expose this configuration to our users.

the discussions around single node concluded that the components should decide what to deploy based on the cluster topology  and that we avoid supporting finer-grained customization from users.

As for handling cases where the topology changes from single node to HA, I don't think it's a concern in the short term.

TL;DR: My gut feeling is that tweaking deployment, statefulset, ... replicas in each `New...()` function is more maintanable/less risky in the short term. We might have to reconsider to support sharding but I also assume that CMO code base will be very different by this time :)

**Comment by romfreiman** [comment]
@lilic @simonpasquier please run e2e-aws-single-node job. If missing, u can add it to this repo. 
Also, u can use cluster bot:
launch 4.8.0-0.nightly aws,single-node,xlarge

**Comment by romfreiman** [comment]
Also, don't forget to remove your component from the sno test:

https://github.com/openshift/origin/blob/master/test/extended/single_node/topology.go#L104

File: pkg/client/client.go [hunk 1]
```diff
@@ -267,6 +268,20 @@ func (c *Client) NamespacesToMonitor() ([]string, error) {
...
 }
 
+func (c *Client) IsHighlyAvaliable() (bool, error) {
+	infra, err := c.GetInfrastructure("cluster")
+	if err != nil {
+		return true, errors.Wrap(err, "could not get Infrastructure 'cluster' instance")
+	}
+	klog.V(6).Infof("control plane topology is: %s", infra.Status.ControlPlaneTopology)
+
+	if infra.Status.ControlPlaneTopology == configv1.SingleReplicaTopologyMode {
```
Comment (wking) [inline pkg/client/client.go:278]: it might be more future-proof to just return `infra.Status.ControlPlaneTopology` [and/or `infra.Status.InfrastructureTopology`][1], instead of collapsing to a boolean here.  That would allow you to `switch` on the value in consuming code, and you could expand those switches later if the API grows more topology modes and monitoring has to do something more than default them to HA.

[1]: https://github.com/openshift/enhancements/pull/671#discussion_r584803866
